cli/sharing.go

```diff
@@ -0,0 +1,218 @@
+package cli
+
+import (
+	"fmt"
+	"regexp"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/cli/cliui"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/serpent"
+)
+
+const defaultGroupDisplay = "-"
+
+type workspaceShareRow struct {
+	User  string                 `table:"user"`
+	Group string                 `table:"group,default_sort"`
+	Role  codersdk.WorkspaceRole `table:"role"`
+}
+
+func (r *RootCmd) sharing() *serpent.Command {
+	orgContext := NewOrganizationContext()
+
+	cmd := &serpent.Command{
+		Use:     "sharing [subcommand]",
+		Short:   "Commands for managing shared workspaces",
+		Aliases: []string{"share"},
+		Handler: func(inv *serpent.Invocation) error {
+			return inv.Command.HelpHandler(inv)
+		},
+		Children: []*serpent.Command{r.shareWorkspace(orgContext)},
+		Hidden:   true,
+	}
+
+	orgContext.AttachOptions(cmd)
+	return cmd
+}
+
+func (r *RootCmd) shareWorkspace(orgContext *OrganizationContext) *serpent.Command {
+	var (
+		// Username regex taken from codersdk/name.go
+		nameRoleRegex = regexp.MustCompile(`(^[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)*)+(?::([A-Za-z0-9-]+))?`)
+		client        = new(codersdk.Client)
+		users         []string
+		groups        []string
+		formatter     = cliui.NewOutputFormatter(
+			cliui.TableFormat(
+				[]workspaceShareRow{}, []string{"User", "Group", "Role"}),
+			cliui.JSONFormat(),
+		)
+	)
+
+	cmd := &serpent.Command{
+		Use:     "add <workspace> --user <user>:<role> --group <group>:<role>",
+		Aliases: []string{"share"},
+		Short:   "Share a workspace with a user or group.",
+		Options: serpent.OptionSet{
+			{
+				Name:        "user",
+				Description: "A comma separated list of users to share the workspace with.",
+				Flag:        "user",
+				Value:       serpent.StringArrayOf(&users),
+			}, {
+				Name:        "group",
+				Description: "A comma separated list of groups to share the workspace with.",
+				Flag:        "group",
+				Value:       serpent.StringArrayOf(&groups),
+			},
+		},
+		Middleware: serpent.Chain(
+			r.InitClient(client),
+			serpent.RequireNArgs(1),
+		),
+		Handler: func(inv *serpent.Invocation) error {
+			if len(users) == 0 && len(groups) == 0 {
+				return xerrors.New("at least one user or group must be provided")
+			}
+
+			workspace, err := namedWorkspace(inv.Context(), client, inv.Args[0])
+			if err != nil {
+				return xerrors.Errorf("could not fetch the workspace %s: %w", inv.Args[0], err)
+			}
+
+			org, err := orgContext.Selected(inv, client)
+			if err != nil {
+				return err
+			}
+
+			userRoles := make(map[string]codersdk.WorkspaceRole, len(users))
+			if len(users) > 0 {
+				orgMembers, err := client.OrganizationMembers(inv.Context(), org.ID)
+				if err != nil {
+					return err
+				}
+
+				for _, user := range users {
+					userAndRole := nameRoleRegex.FindStringSubmatch(user)
```

Comment:
In the event that this is `nil` can we add a check to prevent a panic? 
```suggestion
					userAndRole := nameRoleRegex.FindStringSubmatch(user)
					if userAndRole == nil {
					    return xerrors.Errorf("invalid user format %q: must match pattern 'username:role'", user)
					}
```